pkg/terraform/state/handler.go

```diff
@@ -76,136 +73,178 @@ func (h *handler) OnChange(key string, obj *v1.State) (*v1.State, error) {
+	if runHash == obj.Status.LastRunHash {
+		return obj, nil
+	}
 
-		if obj.Status.ExecutionName != runName {
-			obj.Status.ExecutionName = runName
-			return h.states.Update(obj)
-		}
+	//running an execution
+	obj.Status.LastRunHash = runHash
+	obj, err = h.states.Update(obj)
+	if err != nil {
+		logrus.Debug(err)
+		return obj, nil
+	}
 
-		v1.ExecutionConditionJobDeployed.True(obj)
+	logrus.Debug("lock acquired with new hash")
 
-		return obj, nil
-	})
+	if obj.Spec.Version < 1 {
+		obj.Spec.Version = 1
+	}
+	if obj.Spec.Image == "" {
+		obj.Spec.Image = fmt.Sprintf("%s:latest", DefaultExecutorImage)
+	}
 
+	//new execution if none running
+	exec, err := h.deployCreate(obj, input)
 	if err != nil {
+		logrus.Debug(err)
+
 		return obj, err
 	}
 
+	v1.StateConditionJobDeployed.True(obj)
+	obj.Status.ExecutionName = exec.Name
+
+	//execution was run, find it and set a watcher to clean up when it's done.
+	execution, err := h.executions.Get(obj.Namespace, obj.Status.ExecutionName, metaV1.GetOptions{})
+	if err != nil {
+		return obj, errors.WithMessage(err, "error getting execution")
+	}
+
+	go h.watchExecution(obj, execution, ActionCreate)
+
 	return h.states.Update(obj)
 }
 
 func (h *handler) OnRemove(key string, obj *v1.State) (*v1.State, error) {
 	logrus.Debug("State On Remove Handler")
 	input, ok, err := h.gatherInput(obj)
-	if !ok {
-		v1.ExecutionConditionMissingInfo.True(obj)
-		return h.states.Update(obj)
-	}
 	if err != nil {
+		logrus.Debug("error gathering input")
 		return obj, err
 	}
+	if !ok {
+		v1.ExecutionConditionMissingInfo.SetStatus(obj, err.Error())
+		state, err := h.states.Update(obj)
+		if err != nil {
+			return state, err
+		}
 
-	v1.ExecutionConditionMissingInfo.False(obj)
+		return state, fmt.Errorf("missing info and can not run destroy")
+	}
 
-	if !obj.Spec.DestroyOnDelete {
+	if !obj.Spec.DestroyOnDelete || v1.StateConditionDestroyed.IsTrue(obj) {
 		return obj, nil
 	}
 
-	v1.ExecutionConditionDestroyJobDeployed.False(obj)
+	v1.ExecutionConditionMissingInfo.False(obj)
 
-	var runName string
-	err = v1.ExecutionConditionDestroyJobDeployed.DoUntilTrue(obj, func() (runtime.Object, error) {
-		runName, err = h.deployDestroy(obj, input, ActionDestroy)
-		if err != nil {
-			return obj, err
-		}
+	if v1.StateConditionJobDeployed.IsTrue(obj) && obj.Status.LastRunHash != "" {
+		return obj, fmt.Errorf("job already running %s", obj.Status.LastRunHash)
+	}
 
-		if obj.Status.ExecutionName != runName {
-			obj.Status.ExecutionName = runName
-			obj, err = h.states.Update(obj)
-			if err != nil {
-				return obj, errors.WithMessage(err, "updating execution on state change")
-			}
-		}
-		v1.ExecutionConditionDestroyJobDeployed.True(obj)
-		return h.states.Update(obj)
-	})
+	runHash := createRunHash(obj, input, ActionDestroy)
+	if runHash == obj.Status.LastRunHash {
+		logrus.Debug("hashes the same, nothing to do")
+		return obj, fmt.Errorf("job already running %s", obj.Status.LastRunHash)
```

Comment:
Do you catch this and set the StateConditionJobDeployed condition?

Reply:
Also curious why the error, won't an error cause it to come back through the controller or does OnRemove not work that way in wrangler

Reply:
yes, you want that. everything in OnRemove has to send back an error or it will remove the finalizer and kill the CRD. you can only return nil when for sure the object is ready to be completely removed so you have to just consistently send errors back until the job has completed and run the destroy.